enhancements/authentication/oauth-resource-storage.md

```diff
@@ -0,0 +1,197 @@
+
+## Motivation
+
+Etcd encryption without encrypted oauth access and authorize token is of limited
+value. While these tokens expire after a day usually, the token from a fresh backup
+or the output of etcdctl can be used to act as any oauth authenticated user in the 
+cluster.
+
+A side-effect of the proposed compound tokens is that it allows to hide the secret part
+in the body on LIST or GET requests without returning sensitive information. This
+way users could be allowed to read and possible delete their oauth access tokens, 
+e.g. in order to show that information via oc or console, or to logout certain tokens.
+
+### Goals
+
+- Change the `oauthaccesstokens` and `oauthauthorizetokens` storage format and schema such that
+  the etcd keys alone (which are not encrypted when encryption at REST is enabled) are not sensitive.
+- Allow console and oc to log out all other sessions.
+
+### Non-Goals
+
+## Proposal
+
+We extend the OAuthAccessToken and OAuthAuthorizeToken types in the following way:
+
+```go
+// OAuthAccessToken describes an OAuth access token
+type OAuthAccessToken struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+    ...
+    
+    // SecretToken is the second half of the secret information a user has to pass as bearer token
+    // in order to authorize with the API server. It is concatenated with the name of the 
+    // OAuthAccessToken object.
+    // For old tokens, this field might be empty.
+    // +optional
+    SecretToken string `json:"secretToken,omitempty" protobuf:"bytes,11,opt,name=secretToken"`
+}
+
+// OAuthAuthorizeToken describes an OAuth authorization token
+type OAuthAuthorizeToken struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+    ...
+    
+    // SecretToken is the second half of the secret information a user has to pass as bearer token
+    // in order to authorize with the API server. It is concatenated with the name of the 
+    // OAuthAuthorizeToken object.
+    // For old tokens, this field might be empty.
+    // +optional
+    SecretToken string `json:"secretToken,omitempty" protobuf:"bytes,11,opt,name=secretToken"`
+}
+```
+
+### User Stories [optional]
+
+#### Story 1
+
+As an attacker who got access to an etcd backup I do not get access to oauth tokens.
+
+#### Story 2
+
+As a user I can use the API via oc or console to log out other sessions. 
+
+#### Story 3
+
+(optional) As a user I can list my other oauth sessions, including information like client name and expiration.
+
+### Implementation Details/Notes/Constraints [optional]
+
+The oauth authorizer in kube-apiserver in v4.6 will split the provided bearer token into 
+the first 43 bytes (256 bits, compare https://github.com/openshift/oauth-server/blob/3539e4065084dc0b5f35e43d7ea49631df99760d/pkg/server/crypto/random.go#L31-L35)
```

Comment:
This will break some test cases and potentially some very rarely used features around direct token assignment.  Please call that out.

I'd like to also place a character we expect at byte 44 so that when we change this in the future (and I bet we will), we'll have something clear to look for.

Reply:
No need to obfuscate. `type:<token>` works fine.